clowder-framework/clowder2-frontend
# Datasets V2 CRUD

## Problem
* Primary: verify that Dataset CRUD operations still behave as expected with the V2 API
* Secondary: look into expanding Dataset operations to use the generated V2 client

Fixes #10 

## Approach
Changes:
* Wire up client:
    * Populate `V2.OpenAPI.BASE` with `config.hostname`
    * Populate `V2.OpenAPI.TOKEN` with user's JWT (acquired at login, or from `localStorage`)
* Use new V2 client in Fetch Datasets: `V2.DatasetsService.getDatasetsApiV2DatasetsGet(0, limit)` 
* Use new V2 client in Create Dataset: `V2.DatasetsService.saveDatasetApiV2DatasetsPost(formData)`
* Use new V2 client in Deleted Dataset: `V2.DatasetsService.deleteDatasetApiV2DatasetsDatasetIdDelete(datasetId)`
* Use new V2 client in Get Dataset (About): `V2.DatasetsService.getDatasetApiV2DatasetsDatasetIdGet(id)`
* [WIP] Use new V2 client in Update Dataset. This one requires more work:
    * Current [`PUT /datasets/:id` endpoint](http://clowder.docker.localhost/docs#/datasets/edit_dataset_api_v2_datasets__dataset_id__put) does not accept a `dataset` parameter, so I can't specify a new value
    * Current Datasets view is read-only, so I've added a simple mechanism for editing the dataset's name (mainly for testing, as I don't want to diverge from any existing designs)

## How to Test
### Prerequisites
* run FastAPI backend via Docker Compose: `docker-compose up --build --scale backend=4`
* add `clowder.docker.localhost` to your `/etc/hosts` file
* verify platform is up and running:
```bash
% curl clowder.docker.localhost 
{"status":"ok"}%     
```

### Test Steps
1. Checkout and run this branch locally
2. `export CLOWDER_REMOTE_HOSTNAME="http://clowder.docker.localhost"`
3. Generate V2 client code from running FastAPI backend:`npm run codegen:v2`
4. Start the server, navigate to http://localhost:3000, and log in to the React UI (sign up, if necessary): `npm start`
    * You should see any datasets previously created listed on the Dashboard (Fetch Datasets)
5. On the right, click "Create Dataset"
    * You should see a dialog appear with 2 inputs: name + description
6. Enter a `name` and `description` into the dialog and click "Create"
    * You should see the dialog close and your dataset should be created on the server (Create Dataset)
    * You should be automatically redirected to the Dataset view for the newly-created dataset (Get Datasets)
    * [WIP] You should see an EDIT button beside the dataset's name (on the right side)
7. [WIP] Click the EDIT button
    * [WIP] You should see a text input appear allowing you to edit the dataset's name 
    * [WIP] You should see 2 buttons reading "Save" and "Cancel"
8. [WIP] Click the "Save" button
    * [WIP] You should see the dataset name update on the (Update Dataset)
8. Navigate back to the Dashboard view and Delete a Dataset
    * You should see the dataset that you just deleted disappear from the list on the Dashboard (Delete Dataset)
